### PR TITLE
[Networking] Refinement of message scope; encapsulating the publish logic in the libp2p node

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
### Introduction
This Pull Request (PR) is a continuation of a larger effort to refactor the networking layer structure within the Flow blockchain, specifically focusing on the removal of the `middleware` component. It aligns with the strategic approach outlined in [Issue #6851](https://github.com/dapperlabs/flow-go/issues/6851).

### Objective
The primary goal of this PR is to:
1. **Relocate the Publish Logic:** Move the `Publish` logic of pubsub out of the middleware, placing it within the LibP2P Node, and encapsulating the logic entirely at the LibP2P Node level.
2. **Enhance the Publish Interface:** Refactor the Publish interface of the libp2p node to a higher abstraction level that accepts a single outgoing message scope object, rather than a more tightly coupled interface.

### Design Principles
The following software engineering principles guide the changes:

- **Encapsulation:** By confining the Publish logic to the LibP2P Node level, we are ensuring that the details of the Publish process are hidden from other parts of the system. This promotes modularity and maintainability.
- **Abstraction:** The introduction of a higher-level Publish interface that accepts a single outgoing message scope object simplifies interaction with the publishing system, offering a clearer and less complex method of operation.
- **Atomicity:** Allowing the upper-level networking component to pass the entire message being published as a single atomic component down to the libp2p level ensures transaction integrity and reduces the risk of inconsistencies.
- **Cohesion:** This refactoring strengthens the cohesion within the networking layer, ensuring that related functionalities are kept together, thus making the codebase more understandable and manageable.

### Implementation Details
The specific changes are as follows:
- The `Publish` logic of pubsub has been moved from the middleware to the LibP2P Node.
- The existing Publish interface of the libp2p node has been restructured into a higher-level interface that accepts a single outgoing message scope object.

### Benefits
The resulting abstraction provides a safer and more maintainable structure. By adhering to established software engineering principles, this PR contributes to the overall robustness and efficiency of the networking layer.

### CODE REVIEW

File: network/p2p/test/topic_validator_test.go [hunk 1]
```diff
@@ -260,14 +284,32 @@ func TestTopicValidator_InvalidTopic(t *testing.T) {
...
+	corruptOutgoingMessageScope.On("PayloadType").Return(dummyMessageScope.PayloadType()).Maybe()
+	corruptOutgoingMessageScope.On("Size").Return(dummyMessageScope.Size()).Maybe()
+
+	// create a dummy block proposal to publish from our SN node
+	err = sn2.Publish(timedCtx, corruptOutgoingMessageScope)
 
 	// publish fails because the topic conversion fails
 	require.Error(t, err)
-
+	fmt.Println(err)
```
Comment (kc1116) [inline network/p2p/test/topic_validator_test.go:312]:
```suggestion
	fmt.Println(err)
```

File: network/message/message_scope.go
Comment (gomisha) [inline network/message/message_scope.go:1]: We now will have `network/message/message_scope.go` and `network/message_scope.go` which can be confusing. Can you add some clarity to this file to say how it's different from `network/message_scope.go`?

File: network/p2p/middleware/middleware.go [hunk 2]
```diff
@@ -773,46 +767,9 @@ func (m *Middleware) processMessage(scope *network.IncomingMessageScope) {
...
-
-// IsConnected returns true if this node is connected to the node with id nodeID.
-// All errors returned from this function can be considered benign.
-func (m *Middleware) IsConnected(nodeID flow.Identifier) (bool, error) {
-	peerID, err := m.idTranslator.GetPeerID(nodeID)
-	if err != nil {
-		return false, fmt.Errorf("could not find peer id for target id: %w", err)
-	}
-	return m.libP2PNode.IsConnected(peerID)
+// TODO: publish has made ready to be removed from middleware, and instead the libp2pNode.Publish should be used directly.
```
Comment (gomisha) [inline network/p2p/middleware/middleware.go:770]:
```suggestion
// TODO: DO NOT USE. Publish is ready to be removed from middleware. Use libp2pNode.Publish directly.
```

File: network/p2p/test/sporking_test.go [hunk 3]
```diff
@@ -87,10 +86,12 @@ func TestCrosstalkPreventionOnNetworkKeyChange(t *testing.T) {
 	require.NoError(t, err)
 
 	// create stream from node 1 to node 2
-	testOneToOneMessagingSucceeds(t, node1, peerInfo2)
+	node1.Host().Peerstore().AddAddrs(peerInfo2.ID, peerInfo2.Addrs, peerstore.AddressTTL)
+	s, err := node1.CreateStream(context.Background(), peerInfo2.ID)
+	require.NoError(t, err)
+	assert.NotNil(t, s)
```
Comment (gomisha) [inline network/p2p/test/sporking_test.go:92]: Why remove the encapsulation from `testOneToOneMessagingSucceeds` and have this duplicated inline in 2 places? `testOneToOneMessagingFails` is still used in 2 places.

File: network/p2p/test/topic_validator_test.go [hunk 4]
```diff
@@ -260,14 +284,31 @@ func TestTopicValidator_InvalidTopic(t *testing.T) {
...
+		message.ProtocolTypePubSub)
+	require.NoError(t, err)
+
+	// overrides the topic to be an invalid topic
+	corruptOutgoingMessageScope := mocknetwork.NewOutgoingMessageScope(t)
+	corruptOutgoingMessageScope.On("TargetIds").Return(dummyMessageScope.TargetIds()).Maybe()
+	corruptOutgoingMessageScope.On("Topic").Return(topic).Maybe()
+	corruptOutgoingMessageScope.On("Proto").Return(dummyMessageScope.Proto()).Maybe()
+	corruptOutgoingMessageScope.On("PayloadType").Return(dummyMessageScope.PayloadType()).Maybe()
+	corruptOutgoingMessageScope.On("Size").Return(dummyMessageScope.Size()).Maybe()
```
Comment (gomisha) [inline network/p2p/test/topic_validator_test.go:305]:
```suggestion
	corruptOutgoingMessageScope.On("Topic").Return(topic)
	corruptOutgoingMessageScope.On("Proto").Return(dummyMessageScope.Proto())
	corruptOutgoingMessageScope.On("PayloadType").Return(dummyMessageScope.PayloadType())
	corruptOutgoingMessageScope.On("Size").Return(dummyMessageScope.Size())
```

Comment (gomisha) [reply network/p2p/test/topic_validator_test.go:305]: I ran the test without all the `Maybe()` calls with `-count 10` and it passed, so I think we should remove the `Maybe()` calls which make the test less meaningful if all those calls aren't required.

**Review by gomisha — APPROVED** [review]
LGTM. Please address comments / suggestions before merging.

File: network/message/message_scope.go
Comment (yhassanzadeh13) [reply network/message/message_scope.go:1]: `network/message_scope.go` is the interface definition, and `network/message/message_scope.go` is the implementation of that interface. It is a Golang programming convention to define the interface in the parent package (i.e., `network/`) and implement it in a child package, i.e., `network/message/`

File: network/p2p/test/sporking_test.go (hunk 3 above)
Comment (yhassanzadeh13) [reply network/p2p/test/sporking_test.go:92]: The test helper, comprising roughly five lines of code, served a highly specific function and was utilized in only two locations. Its generic naming did not enhance the abstraction, and the absence of reusability rendered it redundant. Therefore, we eliminated it to make way for code that is more maintainable and easier to comprehend.

File: network/p2p/test/topic_validator_test.go (hunk 4 above)
Comment (yhassanzadeh13) [reply network/p2p/test/topic_validator_test.go]: https://github.com/onflow/flow-go/pull/4635/commits/701e14d8dd940b6da2749925d6c01b858bf3cdd1

**Pull request merged**